### DISCUSSION TITLE
Move script unit test

### DISCUSSION THREAD

**Discussion by keyliaran** [aptos-labs/aptos-developer-discussions#144] [Questions] [answered]
Is there any way to write unit tests for move scripts?

**Comment by kshitijc1506** [comment]
You can use the `#[test]` attribute to mark functions as test cases within a **module**.
```
      #[test]
      fun test_do_something() {
          let resource = do_something(10);
          assert!(resource.value == 10, 42);
      }
```

You can run your tests using the CLI:
`$ aptos move test
`

Here are the [docs](https://aptos.dev/move/book/unit-testing/) on unit tests.

In theory, you could perform unit tests "off-chain" by invoking transactions via scripts. Scripts can call `public entry` and `public` functions within modules. Therefore, your testing coverage may be limited as you can't perform unit tests for private functions.

**Reply by keyliaran** [reply]
I'm seeking guidance on how to create a unit test for a move script. As an example, I have a module named simple_module with two functions: deposit and withdraw. I've developed a script called deposit_and_withdraw that executes both of these functions, and I want to test this script using aptos move test.

Here's the script:
`
script {
    fun deposit_and_withdraw(amount: u64) {
        simple_module::deposit(amount);
        simple_module::withdraw(amount);
    }
}
`
My goal is to write a unit test either within the script directory or to use this script in the module directory for unit tests. Is there a way to achieve this without duplicating the script code into a move module test function?

**Reply by kshitijc1506** [reply, marked as answer]
You **cannot** use `aptos move test` for your `deposit_and_withdraw` script function. Unlike modules, scripts cannot declare structs, friends, or have persistent storage, and they cannot be directly unit tested using the `aptos move test` command. 

You should write your unit tests using the `#[test]` attribute in your modules. You can then call invoke those test functions by using scripts (or just the CLI command). If you _need_ to test the script's logic itself, you would typically do this using a testing framework in your backend (typescript/rust/python), or by observing behavior after deployment to testnet/devnet (this is more integration testing).

**Reply by keyliaran** [reply]
Ok, thx!
